warehouse/oidc/forms/activestate.py

```diff
@@ -0,0 +1,245 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import re
+
+from typing import TypedDict
+
+import requests
+import sentry_sdk
+import wtforms
+
+from warehouse import forms
+from warehouse.i18n import localize as _
+from warehouse.utils.project import PROJECT_NAME_RE
+
+_VALID_ORG_URL_NAME = re.compile(r"^[a-zA-Z0-9-]{3,40}$")
+_VALID_PROJECT_NAME = re.compile(r"^[.a-zA-Z0-9-]{3,40}$")
+
+ACTIVESTATE_GRAPHQL_API_URL = "https://platform.activestate.com/graphql/v1/graphql"
+GRAPHQL_GET_ORGANIZATION = "query($orgname: String) {organizations(where: {display_name: {_eq: $orgname}}) {added}}"  # noqa
+GRAPHQL_GET_ACTOR = (
+    "query($username: String) {users(where: {username: {_eq: $username}}) {user_id}}"
+)
+TIMEOUT = 5
+
+
+class UserResponse(TypedDict):
+    user_id: str
+
+
+class ActiveStatePublisherBase(forms.Form):
+    __params__ = ["organization", "project", "actor"]
+
+    organization = wtforms.StringField(
+        validators=[
+            wtforms.validators.InputRequired(
+                message=_("Specify ActiveState organization name"),
+            ),
+            wtforms.validators.Regexp(
+                _VALID_ORG_URL_NAME,
+                message=_("Invalid ActiveState organization name"),
+            ),
+        ]
+    )
+
+    project = wtforms.StringField(
+        validators=[
+            wtforms.validators.InputRequired(
+                message=_("Specify ActiveState project name")
+            ),
+            wtforms.validators.Regexp(
+                _VALID_PROJECT_NAME,
+                message=_("Invalid ActiveState project name"),
+            ),
+        ]
+    )
+
+    actor = wtforms.StringField(
+        validators=[
+            wtforms.validators.InputRequired(
+                message=("Specify the ActiveState actor username")
+            ),
+            wtforms.validators.Regexp(
+                _VALID_ORG_URL_NAME,
+                message=("Invalid ActiveState username"),
+            ),
+        ]
+    )
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def _lookup_organization(self, org_url_name: str) -> None:
+        """Make gql API call to the ActiveState API to check if the organization
+        exists"""
+        try:
+            response = requests.post(
+                f"{ACTIVESTATE_GRAPHQL_API_URL}",
```

Comment:
Nitpick: unnecessary f-string:

```suggestion
                ACTIVESTATE_GRAPHQL_API_URL,
```

Reply:
Done locally.